### PR TITLE
# 714 extract prevalidation step from transaction sender

### PR DESCRIPTION
**Pull request by rozhkovdmitrii**

### CODE REVIEW

File: proxy/common_neon/emulator_interactor.py [hunk 1]
```diff
@@ -1,21 +1,40 @@
 import json
 import subprocess
 from logged_groups import logged_group
-
```
Comment (rozhkovdmitrii) [inline proxy/common_neon/emulator_interactor.py:4]: roll back

File: proxy/common_neon/emulator_interactor.py [hunk 2]
```diff
@@ -1,21 +1,40 @@
...
+        dst = 'deploy'
+        logger.debug(f'deploy contract: {contract}')
+    else:
+        dst = neon_trx.toAddress.hex()
+        logger.debug(f'destination address {dst}')
+    logger.debug(f"Calling data: {(dst, neon_sender_acc, neon_trx.callData.hex(), hex(neon_trx.value))}")
+    emulator_json = call_emulated(dst, neon_sender_acc, neon_trx.callData.hex(), hex(neon_trx.value))
+    logger.debug(f'emulator returns: {json.dumps(emulator_json, sort_keys=True)}')
+    return emulator_json
+
```
Comment (rozhkovdmitrii) [inline proxy/common_neon/emulator_interactor.py:38]: double space

File: proxy/neon_rpc_api_model/neon_tx_stages.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+import abc
+import os
+
+import base58
+from logged_groups import logged_group
+from solana.publickey import PublicKey
+from solana.transaction import AccountMeta
+
+from proxy.common_neon.address import accountWithSeed
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/neon_tx_stages.py]: ..common_neon

File: proxy/neon_rpc_api_model/operator_resource_list.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    # These variables are global for class, they will be initialized one time
+    _manager = mp.Manager()
+    _free_resource_list = _manager.list()
+    _bad_resource_list = _manager.list()
+    _check_time_resource_list = _manager.list()
+    _resource_list_len = mp.Value(ctypes.c_uint, 0)
+    _last_checked_time = mp.Value(ctypes.c_ulonglong, 0)
+    _resource_list = []
+
+    def __init__(self, neon_tx_sender):
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/operator_resource_list.py]: sender

File: proxy/neon_rpc_api_model/operator_resource_list.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        self._neon_tx_sender = neon_tx_sender
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/operator_resource_list.py]: self._s = sender

File: proxy/neon_rpc_api_model/operator_resource_list.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        return self.get_active_resource()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.free_resource_info()
+
+    @staticmethod
+    def _get_current_time() -> int:
+        return math.ceil(datetime.now().timestamp())
+
+    @staticmethod
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/operator_resource_list.py]: make it back non static

File: proxy/neon_rpc_api_model/operator_resource_list.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+            for idx in range(len(OperatorResourceList._resource_list)):
+                OperatorResourceList._free_resource_list.append(idx)
+                OperatorResourceList._check_time_resource_list.append(0)
+
+            OperatorResourceList._resource_list_len.value = len(OperatorResourceList._resource_list)
+            if OperatorResourceList._resource_list_len.value == 0:
+                raise RuntimeError('Operator has NO resources!')
+
+    def _recheck_bad_resource_list(self) -> int:
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/operator_resource_list.py]: without -> int

File: proxy/neon_rpc_api_model/transaction_sender.py [hunk 8]
```diff
@@ -1,526 +1,70 @@
...
         self.solana = solana
-        self._resource_list = OperatorResourceList(self)
         self.resource = None
         self.signer = None
         self.operator_key = None
         self.builder = None
 
         self._pending_tx = None
 
-        self.eth_sender = '0x' + eth_tx.sender()
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/transaction_sender.py]: rollback renaming

File: proxy/neon_rpc_api_model/transaction_validator.py [hunk 9]
```diff
@@ -11,8 +10,10 @@
 from ..common_neon.solana_interactor import SolanaInteractor
 from ..common_neon.estimate import GasEstimate
 
-from ..environment import ACCOUNT_PERMISSION_UPDATE_INT, CHAIN_ID
-from ..environment import NEON_GAS_LIMIT_MULTIPLIER_NO_CHAINID, ALLOW_UNDERPRICED_TX_WITHOUT_CHAINID
+from ..environment import ACCOUNT_PERMISSION_UPDATE_INT, CHAIN_ID, NEON_GAS_LIMIT_MULTIPLIER_NO_CHAINID, ALLOW_UNDERPRICED_TX_WITHOUT_CHAINID
```
Comment (rozhkovdmitrii) [inline proxy/neon_rpc_api_model/transaction_validator.py]: break after NEON_GAS_LIMIT_MULTIPLIER_NO_CHAINID

**Review by afalaleev — APPROVED** [review]

**Pull request merged**